File: src/parser.py
import re

from src.interfaces import *


def parse_version(log: str) -> PojavLauncherVersion:
    pattern = r'Info: Launcher version: (.*?)-(.*?)-(.*?)-(\S*)'
    results = re.search(pattern, log, re.MULTILINE)
    return PojavLauncherVersion(
        results.group(1), results.group(2),
        results.group(3), results.group(4))


def parse_minecraft_version(log: str) -> MinecraftVersion:
    version_name_pattern = r'Selected Minecraft version: (.*)'
    analyzed_version = MinecraftVersion('UNKNOWN', VersionType.VANILLA)

    regex_version_name = re.search(version_name_pattern, log)
    if regex_version_name:
        analyzed_version.name = regex_version_name.group(1).lower()

    if 'forge' in analyzed_version.name:
        analyzed_version.type = VersionType.FORGE
    if 'optifine' in analyzed_version.name:
        analyzed_version.type = VersionType.OPTIFINE
    if 'fabric' in analyzed_version.name:
        analyzed_version.type = VersionType.FABRIC

    return analyzed_version


def parse_architecture(log: str) -> Architecture:
    pattern = 'Architecture: (.*)$'
    result = re.search(pattern, log, re.MULTILINE).group(1)
    result = result.strip()
    if 'arm64' in result:
        return Architecture.ARM_64
    if 'arm' in result:
        return Architecture.ARM_32

    # TODO detect x86 architecture !
    return Architecture.UNKNOWN


def parse_java_arguments(log: str) -> str:
    pattern = 'Info: Custom Java arguments:.*?\"(.*)\"'
    results = re.search(pattern, log, re.MULTILINE)
    return results.group(1)


def parse_env_variables(log: str) -> dict:
    pattern = 'Added custom env: (.*?)=(.*)$'
    results = re.findall(pattern, log, re.MULTILINE)
    # Build a dict with all the variables
    env_dictionary = {}
    for env_tuple in results:
        env_dictionary[env_tuple[0]] = env_tuple[1]

    return env_dictionary


def parse_java_runtime(java_path: str) -> JavaRuntime:
    java_path = java_path.casefold()
    if java_path.endswith('/internal'):
        return JavaRuntime(8, JavaRuntimeSource.INTERNAL, JavaRuntimeType.JRE)
    if java_path.endswith('/internal17') or java_path.endswith('/internal-17'):
        return JavaRuntime(17, JavaRuntimeSource.INTERNAL, JavaRuntimeType.JRE)

    runtime = JavaRuntime()
    runtime.source = JavaRuntimeSource.EXTERNAL
    if 'jdk' in java_path:
        runtime.type = JavaRuntimeType.JDK
    if 'jre' in java_path:
        runtime.type = JavaRuntimeType.JRE

    pattern = r'(?:jdk|jre)(\d*)'
    results = re.search(pattern, java_path)
    try:
        runtime.version = int(results.group(1))
    except (IndexError, TypeError) as error:
        print('Error during java runtime parsing')

    return runtime


def parse_renderer(log: str) -> PojavRenderer:
    # The pattern is a log line printed by gl4es
    pattern = r'v1\.1\.(\d) built on '
    result = re.search(pattern, log)
    if result:
        gl4es_version = int(result.group(1))
        if gl4es_version == 5:
            return PojavRenderer.GL4ES_115
        if gl4es_version == 4:
            if 'GLSL 320 es supported' in log:
                return PojavRenderer.HOLY_GL4ES_114
            else:
                return PojavRenderer.GL4ES_114

    if 'VirGL: libvirgl_test_server' in log:
        return PojavRenderer.VIRGL

    if 'POJAV_RENDERER=opengles3_desktopgl_angle_vulkan' in log:
        return PojavRenderer.ANGLE

    return PojavRenderer.UNKNOWN


def parse_build_type(log: str) -> BuildType:
    if 'net.kdt.pojavlaunch.debug' in log:
        return BuildType.DEBUG
    if 'net.kdt.pojavlaunch' in log:
        return BuildType.PRODUCTION
    return BuildType.UNKNOWN


def parse_shader_dump(log: str, env_variable: dict) -> list[str]:
    if env_variable.get('LIBGL_VGPU_DUMP'):
        # Note that this method might be quite slow
        # TODO dump original shaders ?
        translated_shaders = re.findall(r'^New VGPU Shader conversion:.*?}\n{2}', log, re.MULTILINE | re.IGNORECASE | re.DOTALL)
        return translated_shaders
    else:
        return []


def parse_minecraft_username(log: str) -> str:
    username = re.search(r'(?<=user): (.*)', log)
    if not username:
        return ""

    return username.group(1)


def parse_forge_mods(content_log: str) -> list[str]:
    """Parse existing mods in the forge error print out: [[state, mod_name]]"""
    return re.findall(r'^\t\| ([ULCHIJADE].*?) {2}\| (.*?) *\|', content_log, re.MULTILINE)

File: parser/parser.go
package parser

import (
	"fmt"
	"strconv"
	"strings"

	regexp "github.com/wasilibs/go-re2"

	"github.com/mathias-boulay/parser/types"
)

/** Build a mc version from */
func ParseVersion(log string) types.PojavLauncherVersion {
	pattern := regexp.MustCompile(`Info: Launcher version: (.*?)-(.*?)-(.*?)-(\S*)`)
	results := pattern.FindStringSubmatch(log)
	return types.PojavLauncherVersion{MajorCode: results[1], CommitNumber: results[2], CommitSha: results[3], Branch: results[4]}
}

/** Build the minecraft parse version */
func ParseMinecraftVersion(log string) types.MinecraftVersion {
	versionNamePattern := regexp.MustCompile(`Selected Minecraft version: (.*)`)
	analyzedVersion := types.MinecraftVersion{Name: "UNKNOWN", Type: types.VANILLA}

	if regexVersionName := versionNamePattern.FindStringSubmatch(log); len(regexVersionName) > 0 {
		analyzedVersion.Name = strings.ToLower(regexVersionName[1])
	}

	types := map[string]types.McVersion{"forge": types.FORGE, "optifine": types.OPTIFINE, "fabric": types.FABRIC, "quilt": types.QUILT}
	for key, mcType := range types {
		if strings.Contains(log, key) {
			analyzedVersion.Type = mcType
			break
		}
	}

	return analyzedVersion
}

/** Parse the mc architecture */
func ParseArchitecture(log string) types.DeviceArchitecture {
	pattern := regexp.MustCompile(`(?m)Architecture: (.*)$`)
	result := pattern.FindStringSubmatch(log)[1]
	result = strings.TrimSpace(result)

	architectures := map[string]types.DeviceArchitecture{"arm64": types.ARM_64, "arm": types.ARM_32, "x86_64": types.X64, "x86": types.X86}
	for key, arch := range architectures {
		if strings.EqualFold(result, key) {
			return arch

		}
	}

	return types.ARCHITECTURE_UNKNOWN
}

/** Get all the env variables */
func ParseEnvVariables(log string) map[string]string {
	pattern := regexp.MustCompile(`(?m)Added custom env: (.*?)=(.*)$`)
	results := pattern.FindAllStringSubmatch(log, -1)
	envDictionary := make(map[string]string)

	for _, envTuple := range results {
		envDictionary[envTuple[1]] = envTuple[2]
	}

	return envDictionary
}

/** Get all the java args */
func ParseJavaArguments(log string) string {
	pattern := regexp.MustCompile(`(?m)Info: Custom Java arguments:.*?"(.*)"`)
	results := pattern.FindStringSubmatch(log)
	return results[1]
}

/** Parse the java runtime */
func ParseJavaRuntime(javaPath string) types.JavaRuntime {
	javaPath = strings.ToLower(javaPath)
	if strings.HasSuffix(javaPath, "/internal") {
		return types.JavaRuntime{Version: 8, Source: types.INTERNAL, Type: types.JRE}
	}
	if strings.HasSuffix(javaPath, "/internal17") || strings.HasSuffix(javaPath, "/internal-17") {
		return types.JavaRuntime{Version: 17, Source: types.INTERNAL, Type: types.JRE}
	}

	var runtime types.JavaRuntime
	runtime.Source = types.EXTERNAL
	if strings.Contains(javaPath, "jdk") {
		runtime.Type = types.JDK
	}
	if strings.Contains(javaPath, "jre") {
		runtime.Type = types.JRE
	}

	pattern := regexp.MustCompile(`(?:jdk|jre)(\d*)`)
	results := pattern.FindStringSubmatch(javaPath)
	if len(results) > 1 {
		version := results[1]
		if parsedVersion, err := strconv.ParseFloat(version, 32); err == nil {
			runtime.Version = float32(parsedVersion)
		} else {
			fmt.Println("Error during java runtime parsing")
		}
	}

	return runtime
}

/** Parse the renderer */
func ParseRenderer(log string) types.Renderer {
	pattern := regexp.MustCompile(`v1\.1\.(\d) built on `)
	result := pattern.FindStringSubmatch(log)
	if len(result) > 1 {
		gl4esVersion, _ := strconv.Atoi(result[1])
		isHoly := strings.Contains(log, "GLSL 320 es supported")
		if gl4esVersion == 5 {
			if isHoly {
				return types.HOLY_GL4ES_115
			} else {
				return types.GL4ES_115
			}
		}
		if gl4esVersion == 4 {
			if isHoly {
				return types.HOLY_GL4ES_114
			} else {
				return types.GL4ES_114
			}
		}
	}

	// TODO support zink, and perhaps vulkanmod as a special case ?
	m := map[string]types.Renderer{"VirGL: libvirgl_test_server": types.VIRGL, "POJAV_RENDERER=opengles3_desktopgl_angle_vulkan": types.ANGLE}
	for key, renderer := range m {
		if strings.Contains(log, key) {
			return renderer
		}
	}

	return types.RENDERER_UNKNOWN
}

/** Gather the build type from the package name */
func ParseBuildType(log string) types.PojavBuildType {
	m := map[string]types.PojavBuildType{"net.kdt.pojavlaunch.debug": types.DEBUG, "net.kdt.pojavlaunch.": types.BUILD_TYPE_UNKNOWN, "net.kdt.pojavlaunch": types.PRODUCTION}
	for key, buildType := range m {
		if strings.Contains(log, key) {
			return buildType
		}
	}

	return types.BUILD_TYPE_UNKNOWN
}

/** Get the mc username */
func ParseMinecraftUsername(log string) string {
	usernamePattern := regexp.MustCompile(`user: (.*)`)
	username := usernamePattern.FindStringSubmatch(log)
	if len(username) > 0 {
		return username[1]
	}
	return ""
}

/** Get all forge mods */
func ParseForgeMods(log string) [][]string {
	modPatterns := regexp.MustCompile(`^\t\| ([ULCHIJADE].*?) {1}\| (.*?) *\|`).FindAllStringSubmatch(log, -1)

	return modPatterns
}

func ParseLog(log string, shouldLocalize bool) types.ParsedLog {
	parsedLog := types.ParsedLog{
		Version:          ParseVersion(log),
		MinecraftVersion: ParseMinecraftVersion(log),
		Renderer:         ParseRenderer(log),
		EnvVariables:     ParseEnvVariables(log),
		Architecture:     ParseArchitecture(log),
		JavaArguments:    ParseJavaArguments(log),
		Username:         ParseMinecraftUsername(log),
	}
	parsedLog.BuildType = ParseBuildType(parsedLog.EnvVariables["JAVA_HOME"])
	parsedLog.JavaRuntime = ParseJavaRuntime(parsedLog.EnvVariables["JAVA_HOME"])

	parsedLog.Errors = CheckErrors(log, parsedLog)
	if shouldLocalize {
		Localize(parsedLog.Errors)
	}

	return parsedLog
}
